flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/FlinkKubeClient.java

```diff
@@ -104,6 +106,67 @@ KubernetesWatch watchPodsAndDoCallback(
 		Map<String, String> labels,
 		WatchCallbackHandler<KubernetesPod> podCallbackHandler);
 
+	/**
+	 * Create the ConfigMap with specified content. If the ConfigMap already exists, a FlinkRuntimeException will be
+	 * thrown.
+	 *
+	 * @param configMap ConfigMap.
+	 *
+	 * @return Return the ConfigMap create future.
+	 */
+	CompletableFuture<Void> createConfigMap(KubernetesConfigMap configMap);
+
+	/**
+	 * Get the ConfigMap with specified name.
+	 *
+	 * @param name ConfigMap name.
+	 *
+	 * @return Return the ConfigMap, or empty if the ConfigMap does not exist.
+	 */
+	Optional<KubernetesConfigMap> getConfigMap(String name);
+
+	/**
+	 * Update an existing ConfigMap with the data. Benefit from <a href=https://kubernetes.io/docs/reference/using-api/api-concepts/#resource-versions>
+	 * resource version</a> and combined with {@link #getConfigMap(String)}, we could perform a get-check-and-update
+	 * transactional operation. Since concurrent modification could happen on a same ConfigMap,
+	 * the update operation may fail. We need to retry internally. The max retry attempts could be
+	 * configured via {@link org.apache.flink.kubernetes.configuration.KubernetesConfigOptions#KUBERNETES_TRANSACTIONAL_OPERATION_MAX_RETRIES}.
+	 *
+	 * @param configMapName ConfigMap to be replaced with.
+	 * @param function      Function to be applied to the obtained ConfigMap and get a new updated one. If the returned
```

Comment:
If we have an exception in the `updateFunction`, we convert it to the unchecked exception and retry again. Since a new retry attempt will get the latest ConfigMap, the `updateFunction` could then execute successfully.

Reply:
What is throwing the exception?  If we only modify the given `ConfigMap`, then I don't see that we have to throw an exception. In any case, this should be added to the JavaDocs.

Reply:
For example, we do the serialize/deserialize/discard-state for the job graph store or completed checkpoint. We could encounter the `Exception`. I think we could retry in such situation.

Reply:
Would we do these kind of calls in the update function of the configMap?

Reply:
Hmm, how would this pattern work with concurrent modifications? Assume that the following happens: We get the `configMap`, discard the `stateHandle` and then return the updated `configMap` w/o the state handle key. Now for some reason `internalClient.createOrReplace` fails and the whole operation will be retried. Now when the `function` is re-executed it will fail right away because `stateHandle` can no longer be discarded but it is still in the `configMap` (because the update failed).

I guess what I want to say is that we should not discard state which has not been removed from the `configMap`. Hence, it would be good if we could support an API where we first update the `configMap` and only do an operation if the update has succeeded.

Reply:
After a careful consideration, I think you are right. We do not need the `FunctionWithException` here. Only the update operation succeeded, we need to do the following external storage cleanup. For example, we could use `kubeClient.checkAndUpdateConfigMap(configMapName, updateFunction).whenCompleteAsync()` for this operation.